# Introduction (from Udacity)
The dataset that you will be wrangling (and analyzing and visualizing) is the tweet archive of Twitter user @dog_rates, also known as WeRateDogs. WeRateDogs is a Twitter account that rates people's dogs with a humorous comment about the dog. These ratings almost always have a denominator of 10. The numerators, though? Almost always greater than 10. 11/10, 12/10, 13/10, etc. Why? Because "they're good dogs Brent." WeRateDogs has over 4 million followers and has received international media coverage.

WeRateDogs downloaded their Twitter archive and sent it to Udacity via email exclusively for you to use in this project. This archive contains basic tweet data (tweet ID, timestamp, text, etc.) for all 5000+ of their tweets as they stood on August 1, 2017. More on this soon.

## Project Motivation
### Context
Your goal: wrangle WeRateDogs Twitter data to create interesting and trustworthy analyses and visualizations. The Twitter archive is great, but it only contains very basic tweet information. Additional gathering, then assessing and cleaning is required for "Wow!"-worthy analyses and visualizations.

## The Data
### Enhanced Twitter Archive

The WeRateDogs Twitter archive contains basic tweet data for all 5000+ of their tweets, but not everything. One column the archive does contain though: each tweet's text, which I used to extract rating, dog name, and dog "stage" (i.e. doggo, floofer, pupper, and puppo) to make this Twitter archive "enhanced." Of the 5000+ tweets, I have filtered for tweets with ratings only (there are 2356).

I extracted this data programmatically, but I didn't do a very good job. The ratings probably aren't all correct. Same goes for the dog names and probably dog stages (see below for more information on these) too. You'll need to assess and clean these columns if you want to use them for analysis and visualization.

### Additional Data via the Twitter API

Back to the basic-ness of Twitter archives: retweet count and favorite count are two of the notable column omissions. Fortunately, this additional data can be gathered by anyone from Twitter's API. Well, "anyone" who has access to data for the 3000 most recent tweets, at least. But you, because you have the WeRateDogs Twitter archive and specifically the tweet IDs within it, can gather this data for all 5000+. And guess what? You're going to query Twitter's API to gather this valuable data.

### Image Predictions File

One more cool thing: I ran every image in the WeRateDogs Twitter archive through a neural network that can classify breeds of dogs*. The results: a table full of image predictions (the top three only) alongside each tweet ID, image URL, and the image number that corresponded to the most confident prediction (numbered 1 to 4 since tweets can have up to four images).

So that's all fun and good. But all of this additional data will need to be gathered, assessed, and cleaned. This is where you come in.

## Key Points
Key points to keep in mind when data wrangling for this project:

- You only want original ratings (**no retweets**) that have **images**. Though there are 5000+ tweets in the dataset, **not all are dog** ratings and some are retweets.
- Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least **8 quality issues** and at least **2 tidiness issues** in this dataset.
- Cleaning includes **merging individual pieces** of data according to the rules of tidy data.
- The fact that the rating numerators are greater than the denominators does not need to be cleaned. This unique rating system is a big part of the popularity of WeRateDogs.
- You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.

## Gathering Data for the Project
Gather each of the three pieces of data as described below in a Jupyter Notebook titled `wrangle_act.ipynb`:

1. The WeRateDogs Twitter archive. I am giving this file to you, so imagine it as a file on hand. Download this file manually by clicking the following link: [`twitter_archive_enhanced.csv`](https://d17h27t6h515a5.cloudfront.net/topher/2017/August/59a4e958_twitter-archive-enhanced/twitter-archive-enhanced.csv)

2. The tweet image predictions, i.e., what breed of dog (or other object, animal, etc.) is present in each tweet according to a neural network. This file (`image_predictions.tsv`) is hosted on Udacity's servers and should be downloaded programmatically using the Requests library and the following URL: https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv

3. Gather each tweet's **retweet count** and **favorite ("like") count** at minimum, and any additional data you find interesting. Using the tweet IDs in the WeRateDogs Twitter archive, 
- query the Twitter API for each tweet's JSON data using Python's [Tweepy](http://www.tweepy.org/) library and 
- store each tweet's entire set of JSON data in a file called `tweet_json.txt` file. 
- Each tweet's JSON data should be written to its own line. 
- Then read this .txt file line by line into a pandas DataFrame with (at minimum) **tweet ID**, **retweet count**, and **favorite count**.

*Note: do not include your Twitter API keys, secrets, and tokens in your project submission.*

In [1]:
# import the necessary libraries
import pandas as pd
import numpy as np
import requests
import tweepy
import json

### 2. Download `image_predictions.tsv` programmatically using the [Requests](https://2.python-requests.org//en/master/) library

In [2]:
image_predictions_url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

In [3]:
response = requests.get(image_predictions_url)
#r = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')

In [4]:
if response.status_code == 200:
    with open("./image-predictions.tsv", "wb") as f:
        f.write(response.content)

In [2]:
img_predictions = pd.read_csv('image-predictions.tsv',sep='\t')#,error_bad_lines=False)

### 3.-6. Gather each tweet's retweet count and favorite ("like") count at minimum, and any additional data you find interesting. Using the tweet IDs in the WeRateDogs Twitter archive

In [6]:
#import tweepy   # already done
#import json     # already done

from tweepy import OAuthHandler 
# Into this we pass our consumer key and secret which was given to us, 
# see http://docs.tweepy.org/en/latest/auth_tutorial.html?highlight=OAuthHandler#oauth-1a-authentication

from timeit import default_timer as timer
# Measure execution time of small code snippets, 
# see https://docs.python.org/3/library/timeit.html

In [22]:
# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions
# !!! The actual keys and tokens must be replaced by 'HIDDEN'
consumer_key = 'HIDDEN' 
consumer_secret = 'HIDDEN' 
access_token = 'HIDDEN' 
access_secret = 'HIDDEN' 

In [23]:
auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

In [24]:
df_1 = pd.read_csv('twitter-archive-enhanced.csv')
df_1.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [25]:
# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API

tweet_ids = df_1.tweet_id.values
len(tweet_ids)

2356

In [27]:
# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print('time taken by the loop:',end - start)
print('fails_dict:',fails_dict)

1: 892420643555336193
Success
2: 892177421306343426
Success
3: 891815181378084864
Success
4: 891689557279858688
Success
5: 891327558926688256
Success
6: 891087950875897856
Success
7: 890971913173991426
Success
8: 890729181411237888
Success
9: 890609185150312448
Success
10: 890240255349198849
Success
11: 890006608113172480
Success
12: 889880896479866881
Success
13: 889665388333682689
Success
14: 889638837579907072
Success
15: 889531135344209921
Success
16: 889278841981685760
Success
17: 888917238123831296
Success
18: 888804989199671297
Success
19: 888554962724278272
Success
20: 888202515573088257
Fail
21: 888078434458587136
Success
22: 887705289381826560
Success
23: 887517139158093824
Success
24: 887473957103951883
Success
25: 887343217045368832
Success
26: 887101392804085760
Success
27: 886983233522544640
Success
28: 886736880519319552
Success
29: 886680336477933568
Success
30: 886366144734445568
Success
31: 886267009285017600
Success
32: 886258384151887873
Success
33: 8860541600590725

Fail
262: 842846295480000512
Success
263: 842765311967449089
Success
264: 842535590457499648
Success
265: 842163532590374912
Success
266: 842115215311396866
Success
267: 841833993020538882
Success
268: 841680585030541313
Success
269: 841439858740625411
Success
270: 841320156043304961
Success
271: 841314665196081154
Success
272: 841077006473256960
Success
273: 840761248237133825
Success
274: 840728873075638272
Success
275: 840698636975636481
Success
276: 840696689258311684
Success
277: 840632337062862849
Success
278: 840370681858686976
Success
279: 840268004936019968
Success
280: 839990271299457024
Success
281: 839549326359670784
Success
282: 839290600511926273
Success
283: 839239871831150596
Success
284: 838952994649550848
Success
285: 838921590096166913
Success
286: 838916489579200512
Success
287: 838831947270979586
Success
288: 838561493054533637
Success
289: 838476387338051585
Success
290: 838201503651401729
Success
291: 838150277551247360
Success
292: 838085839343206401
Success
293

Success
519: 810657578271330305
Success
520: 810284430598270976
Success
521: 810254108431155201
Success
522: 809920764300447744
Success
523: 809808892968534016
Success
524: 809448704142938112
Success
525: 809220051211603969
Success
526: 809084759137812480
Success
527: 808838249661788160
Success
528: 808733504066486276
Success
529: 808501579447930884
Success
530: 808344865868283904
Success
531: 808134635716833280
Success
532: 808106460588765185
Success
533: 808001312164028416
Success
534: 807621403335917568
Success
535: 807106840509214720
Success
536: 807059379405148160
Success
537: 807010152071229440
Success
538: 806629075125202948
Success
539: 806620845233815552
Success
540: 806576416489959424
Success
541: 806542213899489280
Success
542: 806242860592926720
Success
543: 806219024703037440
Success
544: 805958939288408065
Success
545: 805932879469572096
Success
546: 805826884734976000
Success
547: 805823200554876929
Success
548: 805520635690676224
Success
549: 805487436403003392
Success


Success
776: 776201521193218049
Success
777: 776113305656188928
Success
778: 776088319444877312
Success
779: 775898661951791106
Success
780: 775842724423557120
Success
781: 775733305207554048
Success
782: 775729183532220416
Success
783: 775364825476165632
Success
784: 775350846108426240
Success
785: 775096608509886464
Fail
786: 775085132600442880
Success
787: 774757898236878852
Success
788: 774639387460112384
Success
789: 774314403806253056
Success
790: 773985732834758656
Success
791: 773922284943896577
Success
792: 773704687002451968
Success
793: 773670353721753600
Success
794: 773547596996571136
Success
795: 773336787167145985
Success
796: 773308824254029826
Success
797: 773247561583001600
Success
798: 773191612633579521
Success
799: 772877495989305348
Success
800: 772826264096874500
Success
801: 772615324260794368
Success
802: 772581559778025472
Success
803: 772193107915964416
Success
804: 772152991789019136
Success
805: 772117678702071809
Success
806: 772114945936949249
Success
807

Success
1032: 745422732645535745
Success
1033: 745314880350101504
Success
1034: 745074613265149952
Success
1035: 745057283344719872
Success
1036: 744995568523612160
Success
1037: 744971049620602880
Success
1038: 744709971296780288
Success
1039: 744334592493166593
Success
1040: 744234799360020481
Success
1041: 744223424764059648
Success
1042: 743980027717509120
Success
1043: 743895849529389061
Success
1044: 743835915802583040
Success
1045: 743609206067040256
Success
1046: 743595368194129920
Success
1047: 743545585370791937
Success
1048: 743510151680958465
Success
1049: 743253157753532416
Success
1050: 743222593470234624
Success
1051: 743210557239623680
Success
1052: 742534281772302336
Success
1053: 742528092657332225
Success
1054: 742465774154047488
Success
1055: 742423170473463808
Success
1056: 742385895052087300
Success
1057: 742161199639494656
Success
1058: 742150209887731712
Success
1059: 741793263812808706
Success
1060: 741743634094141440
Success
1061: 741438259667034112
Success
10

Success
1281: 708834316713893888
Success
1282: 708810915978854401
Success
1283: 708738143638450176
Success
1284: 708711088997666817
Success
1285: 708479650088034305
Success
1286: 708469915515297792
Success
1287: 708400866336894977
Success
1288: 708356463048204288
Success
1289: 708349470027751425
Success
1290: 708149363256774660
Success
1291: 708130923141795840
Success
1292: 708119489313951744
Success
1293: 708109389455101952
Success
1294: 708026248782585858
Success
1295: 707995814724026368
Success
1296: 707983188426153984
Success
1297: 707969809498152960
Success
1298: 707776935007539200
Success
1299: 707741517457260545
Success
1300: 707738799544082433
Success
1301: 707693576495472641
Success
1302: 707629649552134146
Success
1303: 707610948723478529
Success
1304: 707420581654872064
Success
1305: 707411934438625280
Success
1306: 707387676719185920
Success
1307: 707377100785885184
Success
1308: 707315916783140866
Success
1309: 707297311098011648
Success
1310: 707059547140169728
Success
13

Success
1530: 690248561355657216
Success
1531: 690021994562220032
Success
1532: 690015576308211712
Success
1533: 690005060500217858
Success
1534: 689999384604450816
Success
1535: 689993469801164801
Success
1536: 689977555533848577
Success
1537: 689905486972461056
Success
1538: 689877686181715968
Success
1539: 689835978131935233
Success
1540: 689661964914655233
Success
1541: 689659372465688576
Success
1542: 689623661272240129
Success
1543: 689599056876867584
Success
1544: 689557536375177216
Success
1545: 689517482558820352
Success
1546: 689289219123089408
Success
1547: 689283819090870273
Success
1548: 689280876073582592
Success
1549: 689275259254616065
Success
1550: 689255633275777024
Success
1551: 689154315265683456
Success
1552: 689143371370250240
Success
1553: 688916208532455424
Success
1554: 688908934925697024
Success
1555: 688898160958271489
Success
1556: 688894073864884227
Success
1557: 688828561667567616
Success
1558: 688804835492233216
Success
1559: 688789766343622656
Success
15

Success
1779: 677895101218201600
Success
1780: 677716515794329600
Success
1781: 677700003327029250
Success
1782: 677698403548192770
Success
1783: 677687604918272002
Success
1784: 677673981332312066
Success
1785: 677662372920729601
Success
1786: 677644091929329666
Success
1787: 677573743309385728
Success
1788: 677565715327688705
Success
1789: 677557565589463040
Success
1790: 677547928504967168
Success
1791: 677530072887205888
Success
1792: 677335745548390400
Success
1793: 677334615166730240
Success
1794: 677331501395156992
Success
1795: 677328882937298944
Success
1796: 677314812125323265
Success
1797: 677301033169788928
Success
1798: 677269281705472000
Success
1799: 677228873407442944
Success
1800: 677187300187611136
Success
1801: 676975532580409345
Success
1802: 676957860086095872
Success
1803: 676949632774234114
Success
1804: 676948236477857792
Success
1805: 676946864479084545
Success
1806: 676942428000112642
Success
1807: 676936541936185344
Success
1808: 676916996760600576
Success
18

Success
2028: 671874878652489728
Success
2029: 671866342182637568
Success
2030: 671855973984772097
Success
2031: 671789708968640512
Success
2032: 671768281401958400
Success
2033: 671763349865160704
Success
2034: 671744970634719232
Success
2035: 671743150407421952
Success
2036: 671735591348891648
Success
2037: 671729906628341761
Success
2038: 671561002136281088
Success
2039: 671550332464455680
Success
2040: 671547767500775424
Success
2041: 671544874165002241
Success
2042: 671542985629241344
Success
2043: 671538301157904385
Success
2044: 671536543010570240
Success
2045: 671533943490011136
Success
2046: 671528761649688577
Success
2047: 671520732782923777
Success
2048: 671518598289059840
Success
2049: 671511350426865664
Success
2050: 671504605491109889
Success
2051: 671497587707535361
Success
2052: 671488513339211776
Success
2053: 671486386088865792
Success
2054: 671485057807351808
Success
2055: 671390180817915904
Success
2056: 671362598324076544
Success
2057: 671357843010908160
Success
20

Success
2277: 667443425659232256
Success
2278: 667437278097252352
Success
2279: 667435689202614272
Success
2280: 667405339315146752
Success
2281: 667393430834667520
Success
2282: 667369227918143488
Success
2283: 667211855547486208
Success
2284: 667200525029539841
Success
2285: 667192066997374976
Success
2286: 667188689915760640
Success
2287: 667182792070062081
Success
2288: 667177989038297088
Success
2289: 667176164155375616
Success
2290: 667174963120574464
Success
2291: 667171260800061440
Success
2292: 667165590075940865
Success
2293: 667160273090932737
Success
2294: 667152164079423490
Success
2295: 667138269671505920
Success
2296: 667119796878725120
Success
2297: 667090893657276420
Success
2298: 667073648344346624
Success
2299: 667070482143944705
Success
2300: 667065535570550784
Success
2301: 667062181243039745
Success
2302: 667044094246576128
Success
2303: 667012601033924608
Success
2304: 666996132027977728
Success
2305: 666983947667116034
Success
2306: 666837028449972224
Success
23

In [28]:
len(fails_dict)

23

### 7. Read the json .txt file line by line into a pandas DataFrame with (at minimum) **tweet ID**, **retweet count**, and **favorite count**

In [4]:
# create an empty list 
json_data = []
with open('tweet_json.txt', 'r') as json_file:
    text_line = json_file.readline()
    for text_line in json_file:
        #print(line)
        data_line = json.loads(text_line)
        tweet_id       = data_line['id']
        retweet_count  = data_line['retweet_count']
        favorite_count = data_line['favorite_count']
        #print('tweet_id:',tweet_id,'retweet_count:',retweet_count,'favorite_count:',favorite_count)
        # fill in the list
        json_data.append({'tweet_id':tweet_id,
                          'retweet_count':int(retweet_count),
                          'favorite_count':int(favorite_count)})
# fill in the Pandas Data Frame
tweets_df = pd.DataFrame(json_data, columns = ['tweet_id', 'retweet_count', 'favorite_count'])   

In [5]:
tweets_df.head()

tweet_id  retweet_count  favorite_count
0  892177421306343426           5900           31940
1  891815181378084864           3899           24055
2  891689557279858688           8096           40455
3  891327558926688256           8778           38673
4  891087950875897856           2918           19441

## Assessing Data
- Detect and document at least **eight (8) quality issues** and **two (2) tidiness issues**.
- To meet specifications, the issues that satisfy the Project Motivation (**see the Key Points** header) must be assessed.

### 1) the first file to assess: `twitter-archive-enhanced.csv`

In [6]:
# Convert to Pandas Data Frame
twitter_archive_df = pd.read_csv('twitter-archive-enhanced.csv')

In [7]:
twitter_archive_df.head(50)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0   892420643555336193                    NaN                  NaN   
1   892177421306343426                    NaN                  NaN   
2   891815181378084864                    NaN                  NaN   
3   891689557279858688                    NaN                  NaN   
4   891327558926688256                    NaN                  NaN   
5   891087950875897856                    NaN                  NaN   
6   890971913173991426                    NaN                  NaN   
7   890729181411237888                    NaN                  NaN   
8   890609185150312448                    NaN                  NaN   
9   890240255349198849                    NaN                  NaN   
10  890006608113172480                    NaN                  NaN   
11  889880896479866881                    NaN                  NaN   
12  889665388333682689                    NaN                  NaN   
13  889638837579907072                    NaN                  NaN   
14  889531135344209921                    NaN                  NaN   
15  889278841981685760                    NaN                  NaN   
16  888917238123831296                    NaN                  NaN   
17  888804989199671297                    NaN                  NaN   
18  888554962724278272                    NaN                  NaN   
19  888202515573088257                    NaN                  NaN   
20  888078434458587136                    NaN                  NaN   
21  887705289381826560                    NaN                  NaN   
22  887517139158093824                    NaN                  NaN   
23  887473957103951883                    NaN                  NaN   
24  887343217045368832                    NaN                  NaN   
25  887101392804085760                    NaN                  NaN   
26  886983233522544640                    NaN                  NaN   
27  886736880519319552                    NaN                  NaN   
28  886680336477933568                    NaN                  NaN   
29  886366144734445568                    NaN                  NaN   
30  886267009285017600           8.862664e+17         2.281182e+09   
31  886258384151887873                    NaN                  NaN   
32  886054160059072513                    NaN                  NaN   
33  885984800019947520                    NaN                  NaN   
34  885528943205470208                    NaN                  NaN   
35  885518971528720385                    NaN                  NaN   
36  885311592912609280                    NaN                  NaN   
37  885167619883638784                    NaN                  NaN   
38  884925521741709313                    NaN                  NaN   
39  884876753390489601                    NaN                  NaN   
40  884562892145688576                    NaN                  NaN   
41  884441805382717440                    NaN                  NaN   
42  884247878851493888                    NaN                  NaN   
43  884162670584377345                    NaN                  NaN   
44  883838122936631299                    NaN                  NaN   
45  883482846933004288                    NaN                  NaN   
46  883360690899218434                    NaN                  NaN   
47  883117836046086144                    NaN                  NaN   
48  882992080364220416                    NaN                  NaN   
49  882762694511734784                    NaN                  NaN   

                    timestamp  \
0   2017-08-01 16:23:56 +0000   
1   2017-08-01 00:17:27 +0000   
2   2017-07-31 00:18:03 +0000   
3   2017-07-30 15:58:51 +0000   
4   2017-07-29 16:00:24 +0000   
5   2017-07-29 00:08:17 +0000   
6   2017-07-28 16:27:12 +0000   
7   2017-07-28 00:22:40 +0000   
8   2017-07-27 16:25:51 +0000   
9   2017-07-26 15:59:51 +0000   
10  2017-07-26 00:31:25 +0000   
11  2017-07-25 16:11:53 +0000   
12  2017-07-25

In [8]:
twitter_archive_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [247]:
# Show the duplicates
twitter_archive_df[twitter_archive_df.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [117]:
# Count the sources
twitter_archive_df.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [110]:
# Check the lower case names
#twitter_archive_df[twitter_archive_df.name.str.islower()]
twitter_archive_df[twitter_archive_df.name.str.islower()].name.value_counts()

a               55
the              8
an               7
very             5
just             4
one              4
quite            4
mad              2
actually         2
getting          2
not              2
incredibly       1
space            1
infuriating      1
officially       1
unacceptable     1
life             1
this             1
his              1
my               1
old              1
all              1
by               1
such             1
light            1
Name: name, dtype: int64

### 2) the second file to assess `image_predictions.tsv`

In [16]:
img_predictions.head(20)

tweet_id                                          jpg_url  \
0   666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1   666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2   666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3   666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4   666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5   666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6   666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7   666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8   666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9   666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10  666063827256086533  https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11  666071193221509120  https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12  666073100786774016  https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13  666082916733198337  https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14  666094000022159362  https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15  666099513787052032  https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16  666102155909144576  https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17  666104133288665088  https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18  666268910803644416  https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19  666273097616637952  https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   

    img_num                      p1   p1_conf  p1_dog                  p2  \
0         1  Welsh_springer_spaniel  0.465074    True              collie   
1         1                 redbone  0.506826    True  miniature_pinscher   
2         1         German_shepherd  0.596461    True            malinois   
3         1     Rhodesian_ridgeback  0.408143    True             redbone   
4         1      miniature_pinscher  0.560311    True          Rottweiler   
5         1    Bernese_mountain_dog  0.651137    True    English_springer   
6         1              box_turtle  0.933012   False          mud_turtle   
7         1                    chow  0.692517    True     Tibetan_mastiff   
8         1           shopping_cart  0.962465   False     shopping_basket   
9         1        miniature_poodle  0.201493    True            komondor   
10        1        golden_retriever  0.775930    True     Tibetan_mastiff   
11        1           Gordon_setter  0.503672    True   Yorkshire_terrier   
12        1            Walker_hound  0.260857    True    English_foxhound   
13        1                     pug  0.489814    True        bull_mastiff   
14        1              bloodhound  0.195217    True     German_shepherd   
15        1                   Lhasa  0.582330    True            Shih-Tzu   
16        1          English_setter  0.298617    True        Newfoundland   
17        1                     hen  0.965932   False                cock   
18        1        desktop_computer  0.086502   False                desk   
19        1       Italian_greyhound  0.176053    True         toy_terrier   

     p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0   0.156665    True            Shetland_sheepdog  0.061428    True  
1   0.074192    True          Rhodesian_ridgeback  0.072010    True  
2   0.138584    True                   bloodhound  0.116197    True  
3   0.360687    True           miniature_pinscher  0.222752    True  
4   0.243682    True                     Doberman  0.154629    True  
5   0.263788    True   Greater_Swiss_Mountain_dog  0.016199    True  
6   0.045885   False                     terrapin  0.017885   False  
7   0.058279    True                     fur_coat  0.054449   False  
8   0.014594   False             golden_retriever  0.007959    True  
9   0.192305    True  soft-coated_wheaten_terrier  0.082086    True  
10  0.093718    True           Labrador_retriever 

In [66]:
img_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [122]:
# Check the p1 values
img_predictions.p1.value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
malamute                      30
cocker_spaniel                30
French_bulldog                26
miniature_pinscher            23
Chesapeake_Bay_retriever      23
seat_belt                     22
German_shepherd               20
Siberian_husky                20
Staffordshire_bullterrier     20
Cardigan                      19
web_site                      19
beagle                        18
Shetland_sheepdog             18
teddy                         18
Maltese_dog                   18
Eskimo_dog                    18
Rottweiler                    17
Shih-Tzu                      17
Lakeland_terrier              17
Italian_greyhound             16
kuvasz                        16
          

In [121]:
# Check the number of dogs
img_predictions.p1_dog.value_counts()

True     1532
False     543
Name: p1_dog, dtype: int64

In [118]:
# Chedk the number of images
img_predictions.img_num.value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

In [126]:
# Show the duplicates
img_predictions[img_predictions.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

In [113]:
# Check the tweets with dogs
img_predictions[img_predictions.p1_dog == True]

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
30    666418789513326592    https://pbs.twimg.com/media/CT-YWb7U8AA7QnN.jpg   
31    666421158376562688    https://pbs.twimg.com/media/CT-aggCXAAIMfT3.jpg   
32    666428276349472768    https://pbs.twimg.com/media/CT-g-0DUwAEQdSn.jpg   
34    666435652385423360    https://pbs.twimg.com/media/CT-nsTQWEAEkyDn.jpg   
35    666437273139982337    https://pbs.twimg.com/media/CT-pKmRWIAAxUWj.jpg   
36    666447344410484738    https://pbs.twimg.com/media/CT-yU5QWwAEjLX5.jpg   
37    666454714377183233    https://pbs.twimg.com/media/CT-5Bs-WUAA2JeC.jpg   
38    666644823164719104    https://pbs.twimg.com/media/CUBl6IwVAAA9_zT.jpg   
...                  ...                                                ...   
2041  885311592912609280    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg   
2042  885528943205470208    https://pbs.twimg.com/media/DEoH3yvXgAAzQtS.jpg   
2043  885984800019947520    https://pbs.twimg.com/media/DEumeWWV0AA-Z61.jpg   
2044  886258384151887873    https://pbs.twimg.com/media/DEyfTG4UMAE4aE9.jpg   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsA

In [112]:
# Check the lower case dogs breeds 
img_predictions[(img_predictions.p1_dog == True) & (img_predictions.p1.str.islower())]
#img_predictions[(img_predictions.p1_dog == True) & (img_predictions.p1.str.upper())]

tweet_id                                            jpg_url  \
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
30    666418789513326592    https://pbs.twimg.com/media/CT-YWb7U8AA7QnN.jpg   
36    666447344410484738    https://pbs.twimg.com/media/CT-yU5QWwAEjLX5.jpg   
37    666454714377183233    https://pbs.twimg.com/media/CT-5Bs-WUAA2JeC.jpg   
42    666739327293083650    https://pbs.twimg.com/media/CUC74aTWoAInZey.jpg   
47    666817836334096384    https://pbs.twimg.com/media/CUEDSMEWEAAuXVZ.jpg   
54    667044094246576128    https://pbs.twimg.com/media/CUHREBXXAAE6A9b.jpg   
61    667152164079423490    https://pbs.twimg.com/media/CUIzWk_UwAAfUNq.jpg   
62    667160273090932737    https://pbs.twimg.com/media/CUI6uuaW4AAvCIs.jpg   
63    667165590075940865    https://pbs.twimg.com/media/CUI_kHBWsAAAef5.jpg   
64    667171260800061440    https://pbs.twimg.com/media/CUJEuRIXIAAPDLt.jpg   
65    667174963120574464    https://pbs.twimg.com/media/CUJIFoJWsAAL3Dc.jpg   
66    667176164155375616    https://pbs.twimg.com/media/CUJJLtWWsAE-go5.jpg   
67    667177989038297088    https://pbs.twimg.com/media/CUJK18UWEAEg7AR.jpg   
68    667182792070062081    https://pbs.twimg.com/media/CUJPNjOWsAAZRqP.jpg   
72    667211855547486208    https://pbs.twimg.com/media/CUJppKJWoAA75NP.jpg   
74    667393430834667520    https://pbs.twimg.com/media/CUMOyd3XIAAl13H.jpg   
81    667470559035432960    https://pbs.twimg.com/media/CUNU78YWEAECmpB.jpg   
82    667491009379606528    https://pbs.twimg.com/media/CUNniSlUYAEj1Jl.jpg   
85    667509364010450944    https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg   
...                  ...                                                ...   
1981  871515927908634625    https://pbs.twimg.com/media/DBg_HT9WAAEeIMM.jpg   
1987  872620804844003328    https://pbs.twimg.com/media/DBwr_hzXkAEnZBW.jpg   
1988  872820683541237760    https://pbs.twimg.com/media/DBzhx0PWAAEhl0E.jpg   
1990  873213775632977920    https://pbs.twimg.com/media/DB5HTBGXUAE0TiK.jpg   
1994  874057562936811520    https://pbs.twimg.com/media/DCFGtdoXkAEsqIw.jpg   
1995  874296783580663808    https://pbs.twimg.com/media/DCIgSR0XgAANEOY.jpg   
2001  876484053909872640    https://pbs.twimg.com/media/DCnll_dUQAAkBdG.jpg   
2002  876838120628539392    https://pbs.twimg.com/media/DCsnnZsVwAEfkyi.jpg   
2005  877556246731214848    https://pbs.twimg.com/media/DC20wEcW0AAf59m.jpg   
2009  878281511006478336    https://pbs.twimg.com/media/DDBIX9QVYAAohGa.jpg   
2011  879008229531029506    https://pbs.twimg.com/media/DDLdUrqXYAMOVzY.jpg   
2016  879862464715927552    https://pbs.twimg.com/media/DDXmPrbWAAEKMvy.jpg   
2017  880095782870896641    https://pbs.twimg.com/media/DDa6ckbXgAAM1vV.jpg   
2019  880465832366813184    https://pbs.twimg.com/media/DDgK-J4XUAIEV9W.jpg   
2027  882268110199369728    https://pbs.twimg.com/media/DD5yKdPW0AArzX8.jpg   
2031  883117836046086144    https://pbs.twimg.com/media/DEF2-_hXoAAs62q.jpg   
2032  883360690899218434    https://pbs.twimg.com/media/DEJT3FeXoAAtwUy.jpg   
2033  883482846933004288    https://pbs.twimg.com/media/DELC9dZXUAADqUk.jpg   
2037  884562892145688576    https://pbs.twimg.com/media/DEaZQkfXUA

### 3) the third file to assess `tweet_json.txt`

In [13]:
tweets_df.head(5)

tweet_id  retweet_count  favorite_count
0  892177421306343426           5900           31940
1  891815181378084864           3899           24055
2  891689557279858688           8096           40455
3  891327558926688256           8778           38673
4  891087950875897856           2918           19441

In [14]:
tweets_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2332 entries, 0 to 2331
Data columns (total 3 columns):
tweet_id          2332 non-null int64
retweet_count     2332 non-null int64
favorite_count    2332 non-null int64
dtypes: int64(3)
memory usage: 54.7 KB


In [123]:
# show the duplicates
tweets_df[tweets_df.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, retweet_count, favorite_count]
Index: []

## Quality issues
#### twitter-archive-enhanced.csv
We need only original ratings (no retweets) that have images, however we have

1. replies: 78 , 
2. retweets: 181 , 
3. entries without images = entries without urls = 2356 entries - 2297 expanded_urls = 59.  
4. The columns *in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id	retweeted_status_user_id	retweeted_status_timestamp* are not needed.
5. The *timestamp* is object, while it should be date and time.
6. There are only three sources of images: *Twitter for iPhone*, *Vine - Make a Scene*, *Twitter Web Client*, and *TweetDeck*, which is not seen, because the full url is given.
7. Lower case names are not names

#### img_predictions
8. p1, p2, and p3 column names are unclear
9. The 543 images are not dogs
10. The 619 breed names are lower case, but 1532-692 = 840 are upper case 
11. The underscore in the breed names is unnecessary

### Tidiness issues
1. Merge all the tables into the new one on 'tweet_id' index 

#### twitter-archive-enhanced.csv
2. The dog "stage" (i.e. doggo, floofer, pupper, and puppo) is a variable. The 'stage' should be one column.

## Cleaning

In [246]:
# Copy the existing tables to the new ones
twitter_archive_clean = twitter_archive_df
img_predictions_clean = img_predictions
tweets_clean = tweets_df

### Define
- Quality issues 1, 2, and 3

**Delete all replies, retweets, and entries without images from `twitter_archive_df`** 

### Code

In [346]:
# Fill in the dataframe
twitter_archive_clean = twitter_archive_df[(twitter_archive_df.in_reply_to_user_id.notnull()==False) &
                                           (twitter_archive_df.retweeted_status_user_id.notnull()==False) &
                                           (twitter_archive_df.expanded_urls.notnull()==True) ]

### Test

In [347]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2094 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2094 non-null object
source                        2094 non-null object
text                          2094 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2094 non-null object
rating_numerator              2094 non-null int64
rating_denominator            2094 non-null int64
name                          2094 non-null object
doggo                         2094 non-null object
floofer                       2094 non-null object
pupper                        2094 non-null object
puppo                         2094 non-null object
dtypes: float64(4), int64(3), object(10)

### Define
- Quality issue 4. The columns *in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id	retweeted_status_user_id	retweeted_status_timestamp* are not needed.

**Remove the columns**

### Code

In [348]:
twitter_archive_clean = twitter_archive_clean.drop(['in_reply_to_status_id', 
                                                    'in_reply_to_user_id', 
                                                    'retweeted_status_id',
                                                    'retweeted_status_user_id',
                                                    'retweeted_status_timestamp'
                                                   ], axis=1)

### Test

In [349]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2094 non-null int64
timestamp             2094 non-null object
source                2094 non-null object
text                  2094 non-null object
expanded_urls         2094 non-null object
rating_numerator      2094 non-null int64
rating_denominator    2094 non-null int64
name                  2094 non-null object
doggo                 2094 non-null object
floofer               2094 non-null object
pupper                2094 non-null object
puppo                 2094 non-null object
dtypes: int64(3), object(9)
memory usage: 212.7+ KB


### Define
- Quality issue 5

**The *timestamp* is object, while it should be date and time.**


### Code

In [350]:
# Convert 'timestamp' to datetime
twitter_archive_clean.timestamp = pd.to_datetime(twitter_archive_clean.timestamp)

### Test

In [351]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2094 non-null int64
timestamp             2094 non-null datetime64[ns, UTC]
source                2094 non-null object
text                  2094 non-null object
expanded_urls         2094 non-null object
rating_numerator      2094 non-null int64
rating_denominator    2094 non-null int64
name                  2094 non-null object
doggo                 2094 non-null object
floofer               2094 non-null object
pupper                2094 non-null object
puppo                 2094 non-null object
dtypes: datetime64[ns, UTC](1), int64(3), object(8)
memory usage: 212.7+ KB


### Define
- Quality issue 6. There are only three sources of images: *Twitter for iPhone*, *Vine - Make a Scene*, *Twitter Web Client*, and *TweetDeck*, which is not seen, because the full url is given.

**Replace the source names**

### Code

In [352]:
twitter_archive_clean.loc[twitter_archive_clean.source == '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>', 'source'] = 'Twitter for iPhone'
twitter_archive_clean.loc[twitter_archive_clean.source == '<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>', 'source'] = 'Vine - Make a Scene'
twitter_archive_clean.loc[twitter_archive_clean.source == '<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>', 'source'] = 'Twitter Web Client'
twitter_archive_clean.loc[twitter_archive_clean.source == '<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>', 'source'] = 'TweetDeck'

### Test

In [353]:
twitter_archive_clean.source.value_counts()

Twitter for iPhone     1962
Vine - Make a Scene      91
Twitter Web Client       30
TweetDeck                11
Name: source, dtype: int64

In [354]:
twitter_archive_clean

tweet_id                 timestamp              source  \
0     892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1     892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2     891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3     891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4     891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   
5     891087950875897856 2017-07-29 00:08:17+00:00  Twitter for iPhone   
6     890971913173991426 2017-07-28 16:27:12+00:00  Twitter for iPhone   
7     890729181411237888 2017-07-28 00:22:40+00:00  Twitter for iPhone   
8     890609185150312448 2017-07-27 16:25:51+00:00  Twitter for iPhone   
9     890240255349198849 2017-07-26 15:59:51+00:00  Twitter for iPhone   
10    890006608113172480 2017-07-26 00:31:25+00:00  Twitter for iPhone   
11    889880896479866881 2017-07-25 16:11:53+00:00  Twitter for iPhone   
12    889665388333682689 2017-07-25 01:55:32+00:00  Twitter for iPhone   
13    889638837579907072 2017-07-25 00:10:02+00:00  Twitter for iPhone   
14    889531135344209921 2017-07-24 17:02:04+00:00  Twitter for iPhone   
15    889278841981685760 2017-07-24 00:19:32+00:00  Twitter for iPhone   
16    888917238123831296 2017-07-23 00:22:39+00:00  Twitter for iPhone   
17    888804989199671297 2017-07-22 16:56:37+00:00  Twitter for iPhone   
18    888554962724278272 2017-07-22 00:23:06+00:00  Twitter for iPhone   
20    888078434458587136 2017-07-20 16:49:33+00:00  Twitter for iPhone   
21    887705289381826560 2017-07-19 16:06:48+00:00  Twitter for iPhone   
22    887517139158093824 2017-07-19 03:39:09+00:00  Twitter for iPhone   
23    887473957103951883 2017-07-19 00:47:34+00:00  Twitter for iPhone   
24    887343217045368832 2017-07-18 16:08:03+00:00  Twitter for iPhone   
25    887101392804085760 2017-07-18 00:07:08+00:00  Twitter for iPhone   
26    886983233522544640 2017-07-17 16:17:36+00:00  Twitter for iPhone   
27    886736880519319552 2017-07-16 23:58:41+00:00  Twitter for iPhone   
28    886680336477933568 2017-07-16 20:14:00+00:00  Twitter for iPhone   
29    886366144734445568 2017-07-15 23:25:31+00:00  Twitter for iPhone   
31    886258384151887873 2017-07-15 16:17:19+00:00  Twitter for iPhone   
...                  ...                       ...                 ...   
2326  666411507551481857 2015-11-17 00:24:19+00:00  Twitter for iPhone   
2327  666407126856765440 2015-11-17 00:06:54+00:00  Twitter for iPhone   
2328  666396247373291520 2015-11-16 23:23:41+00:00  Twitter for iPhone   
2329  666373753744588802 2015-11-16 21:54:18+00:00  Twitter for iPhone   
2330  666362758909284353 2015-11-16 21:10:36+00:00  Twitter for iPhone   
2331  666353288456101888 2015-11-16 20:32:58+00:00  Twitter for iPhone   
2332  666345417576210432 2015-11-16 20:01:42+00:00  Twitter for iPhone   
2333  666337882303524864 2015-11-16 19:31:45+00:00  Twitter for iPhone   
2334  666293911632134144 2015-11-16 16:37:02+00:00  Twitter for iPhone   
2335  666287406224695296 2015-11-16 16:11:11+00:00  Twitter for iPhone   
2336  666273097616637952 2015-11-16 15:14:19+00:00  Twitter for iPhone   
2337  666268910803644416 2015-11-16 14:57:41+00:00  Twitter for iPhone   
2338  666104133288665088 2015-11-16 04:02:55+00:00  Twitter for iPhone   
2339  666102155909144576 2015-11-16 03:55:04+00:00  Twitter for iPhone   
2340  666099513787052032 2015-11-16 03:44:34+00:00  Twitter for iPhone   
2341  666094000022159362 2015-11-16 03:22:39+00:00  Twitter for iPhone   
2342  666082916733198337 2015-11-16 02:38:37+00:00  Twitter for iPhone   
2343  666073100786774016 2015-11-16 01:59:36+00:00  Twitter for iPhone   
2344  666071193221509120 2015-11-16 01:52:02+00:00  Twitter for iPhone   
2345  666063827256086533 2015-11-16 01:22:45+00:00  Twitter for iPhone   
2346  666058600524156928 2015-11-16 01:01:59+00:00  Twitter for iPhone   
2347  666057090499244032 2015-11-16 00:55:59+00:00  Twitter for iPhone   
2348  66605552504240

### Define
- Quality issue 7. Lower case names are not names.

**Remove all lower case 'names'**

### Code

In [355]:
# Get the nonames
noname = list(twitter_archive_clean[twitter_archive_clean.name.str.islower()].name.unique())
noname

['such',
 'a',
 'quite',
 'not',
 'one',
 'incredibly',
 'very',
 'my',
 'his',
 'an',
 'actually',
 'just',
 'getting',
 'mad',
 'this',
 'unacceptable',
 'all',
 'old',
 'infuriating',
 'the',
 'by',
 'officially',
 'life',
 'light',
 'space']

In [356]:
for n in noname:
#    print(n)
    twitter_archive_clean.loc[twitter_archive_clean.name == n, 'name'] = ''
#        print('1')
        #print(twitter_archive_clean.tweet_id)
#df_tweets_clean.loc[df_tweets_clean.name == nonname, 'name'] = None

### Test

In [357]:
twitter_archive_clean[twitter_archive_clean.name.str.islower()].name.unique()

array([], dtype=object)

### Define
#### Tidiness issue 1
* Merge all the tables into the new one on *tweet_id* index

### Code

In [220]:
# Delete the dataframe
del we_rate_dogs_clean

In [221]:
# Merge twitter_archive_df with img_predictions
we_rate_dogs_clean = pd.merge(twitter_archive_df, img_predictions, on='tweet_id', how='inner')
# Merge the result with tweets_df
we_rate_dogs_clean = pd.merge(we_rate_dogs_clean, tweets_df, on='tweet_id', how='inner')

### Test

In [222]:
we_rate_dogs_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892177421306343426                    NaN                  NaN   
1     891815181378084864                    NaN                  NaN   
2     891689557279858688                    NaN                  NaN   
3     891327558926688256                    NaN                  NaN   
4     891087950875897856                    NaN                  NaN   
5     890971913173991426                    NaN                  NaN   
6     890729181411237888                    NaN                  NaN   
7     890609185150312448                    NaN                  NaN   
8     890240255349198849                    NaN                  NaN   
9     890006608113172480                    NaN                  NaN   
10    889880896479866881                    NaN                  NaN   
11    889665388333682689                    NaN                  NaN   
12    889638837579907072                    NaN                  NaN   
13    889531135344209921                    NaN                  NaN   
14    889278841981685760                    NaN                  NaN   
15    888917238123831296                    NaN                  NaN   
16    888804989199671297                    NaN                  NaN   
17    888554962724278272                    NaN                  NaN   
18    888078434458587136                    NaN                  NaN   
19    887705289381826560                    NaN                  NaN   
20    887517139158093824                    NaN                  NaN   
21    887473957103951883                    NaN                  NaN   
22    887343217045368832                    NaN                  NaN   
23    887101392804085760                    NaN                  NaN   
24    886983233522544640                    NaN                  NaN   
25    886736880519319552                    NaN                  NaN   
26    886680336477933568                    NaN                  NaN   
27    886366144734445568                    NaN                  NaN   
28    886258384151887873                    NaN                  NaN   
29    885984800019947520                    NaN                  NaN   
...                  ...                    ...                  ...   
2030  666411507551481857                    NaN                  NaN   
2031  666407126856765440                    NaN                  NaN   
2032  666396247373291520                    NaN                  NaN   
2033  666373753744588802                    NaN                  NaN   
2034  666362758909284353                    NaN                  NaN   
2035  666353288456101888                    NaN                  NaN   
2036  666345417576210432                    NaN                  NaN   
2037  666337882303524864                    NaN                  NaN   
2038  666293911632134144                    NaN                  NaN   
2039  666287406224695296                    NaN                  NaN   
2040  666273097616637952                    NaN                  NaN   
2041  666268910803644416                    NaN                  NaN   
2042  666104133288665088                    NaN                  NaN   
2043  666102155909144576                    NaN                  NaN   
2044  666099513787052032                    NaN                  NaN   
2045  666094000022159362                    NaN                  NaN   
2046  666082916733198337                    NaN                  NaN   
2047  666073100786774016                    NaN                  NaN   
2048  666071193221509120                    NaN                  NaN   
2049  666063827256086533                    NaN                  NaN   
2050  666058600524156928                    NaN                  NaN   
2051  666057090499244032                    NaN                  NaN   
2052  666055525042405380                    NaN                  NaN   
2053  666051853826850816                    NaN         

In [198]:
we_rate_dogs_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2060 entries, 0 to 2059
Data columns (total 30 columns):
tweet_id                      2060 non-null int64
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2060 non-null object
source                        2060 non-null object
text                          2060 non-null object
retweeted_status_id           74 non-null float64
retweeted_status_user_id      74 non-null float64
retweeted_status_timestamp    74 non-null object
expanded_urls                 2060 non-null object
rating_numerator              2060 non-null int64
rating_denominator            2060 non-null int64
name                          2060 non-null object
doggo                         2060 non-null object
floofer                       2060 non-null object
pupper                        2060 non-null object
puppo                         2060 non-null object
jpg_url                       2060 

In [223]:
# Show duplicate tweet_id
we_rate_dogs_clean[we_rate_dogs_clean.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog, retweet_count, favorite_count]
Index: []

[0 rows x 30 columns]

### Define
#### Tidiness issue 2
- The dog "stage" (i.e. doggo, floofer, pupper, and puppo) is a variable. The 'stage' should be one column.

### Code

In [228]:
# Get column names
columns = list(we_rate_dogs_clean.columns.values)
# Remove 'doggo', 'floofer', 'pupper', 'puppo' from the list
stages = {'doggo', 'floofer', 'pupper', 'puppo'}
columns = [e for e in columns if e not in stages]
columns

['tweet_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'jpg_url',
 'img_num',
 'p1',
 'p1_conf',
 'p1_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog',
 'retweet_count',
 'favorite_count']

In [229]:
we_rate_dogs_clean = pd.melt(we_rate_dogs_clean, id_vars=columns, value_vars=stages,
                           var_name='stages', value_name='dog_stage')

In [230]:
we_rate_dogs_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8240 entries, 0 to 8239
Data columns (total 23 columns):
tweet_id              8240 non-null int64
timestamp             8240 non-null datetime64[ns, UTC]
source                8240 non-null object
text                  8240 non-null object
expanded_urls         8240 non-null object
rating_numerator      8240 non-null int64
rating_denominator    8240 non-null int64
name                  8240 non-null object
jpg_url               8240 non-null object
img_num               8240 non-null int64
p1                    8240 non-null object
p1_conf               8240 non-null float64
p1_dog                8240 non-null bool
p2                    8240 non-null object
p2_conf               8240 non-null float64
p2_dog                8240 non-null bool
p3                    8240 non-null object
p3_conf               8240 non-null float64
p3_dog                8240 non-null bool
retweet_count         8240 non-null int64
favorite_count        8240 n

In [231]:
we_rate_dogs_clean.stages.value_counts()

floofer    2060
puppo      2060
doggo      2060
pupper     2060
Name: stages, dtype: int64

In [232]:
we_rate_dogs_clean.dog_stage.value_counts()

None       7908
pupper      222
doggo        78
puppo        24
floofer       8
Name: dog_stage, dtype: int64

In [233]:
we_rate_dogs_clean.drop_duplicates(inplace=True)

In [234]:
we_rate_dogs_clean.stages.value_counts()

floofer    2060
puppo      2060
doggo      2060
pupper     2060
Name: stages, dtype: int64

In [245]:
we_rate_dogs_clean.dog_stage.value_counts()

None       2060
pupper      222
doggo        78
puppo        24
floofer       8
Name: dog_stage, dtype: int64

In [236]:
we_rate_dogs_clean

tweet_id                 timestamp  \
0     892177421306343426 2017-08-01 00:17:27+00:00   
1     891815181378084864 2017-07-31 00:18:03+00:00   
2     891689557279858688 2017-07-30 15:58:51+00:00   
3     891327558926688256 2017-07-29 16:00:24+00:00   
4     891087950875897856 2017-07-29 00:08:17+00:00   
5     890971913173991426 2017-07-28 16:27:12+00:00   
6     890729181411237888 2017-07-28 00:22:40+00:00   
7     890609185150312448 2017-07-27 16:25:51+00:00   
8     890240255349198849 2017-07-26 15:59:51+00:00   
9     890006608113172480 2017-07-26 00:31:25+00:00   
10    889880896479866881 2017-07-25 16:11:53+00:00   
11    889665388333682689 2017-07-25 01:55:32+00:00   
12    889638837579907072 2017-07-25 00:10:02+00:00   
13    889531135344209921 2017-07-24 17:02:04+00:00   
14    889278841981685760 2017-07-24 00:19:32+00:00   
15    888917238123831296 2017-07-23 00:22:39+00:00   
16    888804989199671297 2017-07-22 16:56:37+00:00   
17    888554962724278272 2017-07-22 00:23:06+00:00   
18    888078434458587136 2017-07-20 16:49:33+00:00   
19    887705289381826560 2017-07-19 16:06:48+00:00   
20    887517139158093824 2017-07-19 03:39:09+00:00   
21    887473957103951883 2017-07-19 00:47:34+00:00   
22    887343217045368832 2017-07-18 16:08:03+00:00   
23    887101392804085760 2017-07-18 00:07:08+00:00   
24    886983233522544640 2017-07-17 16:17:36+00:00   
25    886736880519319552 2017-07-16 23:58:41+00:00   
26    886680336477933568 2017-07-16 20:14:00+00:00   
27    886366144734445568 2017-07-15 23:25:31+00:00   
28    886258384151887873 2017-07-15 16:17:19+00:00   
29    885984800019947520 2017-07-14 22:10:11+00:00   
...                  ...                       ...   
8210  666411507551481857 2015-11-17 00:24:19+00:00   
8211  666407126856765440 2015-11-17 00:06:54+00:00   
8212  666396247373291520 2015-11-16 23:23:41+00:00   
8213  666373753744588802 2015-11-16 21:54:18+00:00   
8214  666362758909284353 2015-11-16 21:10:36+00:00   
8215  666353288456101888 2015-11-16 20:32:58+00:00   
8216  666345417576210432 2015-11-16 20:01:42+00:00   
8217  666337882303524864 2015-11-16 19:31:45+00:00   
8218  666293911632134144 2015-11-16 16:37:02+00:00   
8219  666287406224695296 2015-11-16 16:11:11+00:00   
8220  666273097616637952 2015-11-16 15:14:19+00:00   
8221  666268910803644416 2015-11-16 14:57:41+00:00   
8222  666104133288665088 2015-11-16 04:02:55+00:00   
8223  666102155909144576 2015-11-16 03:55:04+00:00   
8224  666099513787052032 2015-11-16 03:44:34+00:00   
8225  666094000022159362 2015-11-16 03:22:39+00:00   
8226  666082916733198337 2015-11-16 02:38:37+00:00   
8227  666073100786774016 2015-11-16 01:59:36+00:00   
8228  666071193221509120 2015-11-16 01:52:02+00:00   
8229  666063827256086533 2015-11-16 01:22:45+00:00   
8230  666058600524156928 2015-11-16 01:01:59+00:00   
8231  666057090499244032 2015-11-16 00:55:59+00:00   
8232  666055525042405380 2015-11-16 00:49:46+00:00   
8233  666051853826850816 2015-11-16 00:35:11+00:00   
8234  666050758794694657 2015-11-16 00:30:50+00:00   
8235  666049248165822465 2015-11-16 00:24:50+00:00   
8236  666044226329800704 2015-11-16 00:04:52+00:00   
8237  666033412701032449 2015-11-15 23:21:54+00:00   
8238  666029285002620928 2015-11-15 23:05:30+00:00   
8239  666020888022790149 2015-11-15 22:32:08+00:00   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
5     <a href="http://twitter.com/download/iphone" r...   
6     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
8     <a href="http://twitter.com/download/iphone" r...   
9     <a href="http://twitter.com/download/iphone" r...   
10    <a href="htt

In [237]:
we_rate_dogs_clean = we_rate_dogs_clean.drop('stages',axis=1)

In [243]:
we_rate_dogs_clean.tweet_id.drop_duplicates(inplace=True)

In [244]:
we_rate_dogs_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2392 entries, 0 to 7057
Data columns (total 22 columns):
tweet_id              2392 non-null int64
timestamp             2392 non-null datetime64[ns, UTC]
source                2392 non-null object
text                  2392 non-null object
expanded_urls         2392 non-null object
rating_numerator      2392 non-null int64
rating_denominator    2392 non-null int64
name                  2392 non-null object
jpg_url               2392 non-null object
img_num               2392 non-null int64
p1                    2392 non-null object
p1_conf               2392 non-null float64
p1_dog                2392 non-null bool
p2                    2392 non-null object
p2_conf               2392 non-null float64
p2_dog                2392 non-null bool
p3                    2392 non-null object
p3_conf               2392 non-null float64
p3_dog                2392 non-null bool
retweet_count         2392 non-null int64
favorite_count        2392 n